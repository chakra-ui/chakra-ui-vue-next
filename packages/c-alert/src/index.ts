import { defineComponent, computed, createVNode, mergeProps, isVNode, Fragment } from 'vue';
import { createStylesContext, useMultiStyleConfig, chakra } from '@chakra-ui/vue-system';
import { createContext, getValidChildren } from '@chakra-ui/vue-utils';
import { createIconComponent } from '@chakra-ui/c-icon';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var CInfoIcon = createIconComponent("info");
var CCheckIcon = createIconComponent("check-circle");
var CWarningIcon = createIconComponent("warning-alt");
var CErrorIcon = createIconComponent("warning");

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var STATUSES = {
  info: {
    colorScheme: "blue",
    icon: CInfoIcon
  },
  success: {
    colorScheme: "green",
    icon: CCheckIcon
  },
  warning: {
    colorScheme: "orange",
    icon: CWarningIcon
  },
  error: {
    colorScheme: "red",
    icon: CErrorIcon
  },
  loading: {
    icon: CInfoIcon,
    colorScheme: "blue"
  }
};

var _createStylesContext = createStylesContext("Alert"),
    StylesProvider = _createStylesContext[0],
    useStyles = _createStylesContext[1];

var _createContext = createContext({
  name: "AlertContext",
  errorMessage: "useAlertContext: `context` is undefined. Seems you forgot to wrap alert components in `<c-alert />`"
}),
    AlertProvider = _createContext[0],
    useAlertContext = _createContext[1];
/**
 * CAlert component
 *
 * This is the container component for all Alert components.
 * It also provides state and context to it's compound components
 */


var CAlert = defineComponent({
  name: "CAlert",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    },
    status: {
      type: [String],
      "default": "info"
    },
    colorScheme: {
      type: [String]
    },
    styleConfig: {
      type: [Object]
    },
    variant: {
      type: [String],
      "default": "solid"
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var colorScheme = computed(function () {
      var _STATUSES$props$statu;

      return props.colorScheme || ((_STATUSES$props$statu = STATUSES[props == null ? void 0 : props.status]) == null ? void 0 : _STATUSES$props$statu.colorScheme);
    });
    var themingProps = computed(function () {
      return {
        colorScheme: colorScheme.value,
        variant: props.variant
      };
    });
    AlertProvider({
      status: computed(function () {
        return props.status;
      })
    });
    var styles = useMultiStyleConfig("Alert", themingProps);
    StylesProvider(styles);
    var alertStyles = computed(function () {
      return _extends({
        width: "100%",
        display: "flex",
        alignItems: "center",
        position: "relative",
        overflow: "hidden"
      }, styles.value.container);
    });
    return function () {
      return createVNode(chakra.div, mergeProps({
        "role": "alert",
        "__label": "alert",
        "__css": alertStyles.value
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertTitle component
 *
 * The title component for alerts
 */

var CAlertTitle = defineComponent({
  name: "CAlertTitle",
  setup: function setup(_, _ref2) {
    var attrs = _ref2.attrs,
        slots = _ref2.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__title",
        "__css": styles.value.title
      }, attrs), _isSlot(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});
/**Â«
 * CAlertDescription component
 *
 * The description component for alerts
 */

var CAlertDescription = defineComponent({
  name: "CAlertDescription",
  setup: function setup(_, _ref3) {
    var attrs = _ref3.attrs,
        slots = _ref3.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__description",
        "__css": styles.value.description
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertIcon component
 *
 * The Icon component for alerts
 */

var CAlertIcon = defineComponent({
  name: "CAlertIcon",
  setup: function setup(_, _ref4) {
    var attrs = _ref4.attrs,
        slots = _ref4.slots;
    var styles = useStyles();

    var _useAlertContext = useAlertContext(),
        status = _useAlertContext.status;

    var BaseIcon = STATUSES[status.value].icon;
    var css = computed(function () {
      return status.value === "loading" ? styles.value.spinner : styles.value.icon;
    });
    return function () {
      var validChildren = getValidChildren(slots);
      return createVNode(chakra.span, mergeProps({
        "display": "inherit",
        "__label": "alert__icon"
      }, attrs, {
        "__css": css.value
      }), {
        "default": function _default() {
          return createVNode(Fragment, null, [validChildren.length ? slots : createVNode(BaseIcon, {
            "h": "100%",
            "w": "100%"
          }, null)]);
        }
      }); // return <icon {...styles.value.icon} {...attrs}></icon>
    };
  }
});

export { CAlert, CAlertDescription, CAlertIcon, CAlertTitle };
