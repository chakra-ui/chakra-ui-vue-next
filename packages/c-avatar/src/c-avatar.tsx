/**
 * Hey! Welcome to @chakra-ui/vue-next CAvatar
 *
 * The avatar component is used to represent user and displays the profile picture initials or fallback icon
 *
 * @see Docs     https://next.vue.chakra-ui.com/c-avatar
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue-next/blob/main/packages/c-avatar/src/c-avatar/c-avatar.ts
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2
 */

import {
  h,
  defineComponent,
  PropType,
  computed,
  Fragment,
  ref,
  ComputedRef,
  VNode,
  DefineComponent,
} from "vue"
import { match, SNAO, useThemingProps } from "@chakra-ui/vue-utils"
import { createContext, vueThemingProps } from "@chakra-ui/vue-utils"
import { mergeWith, filterUndefined, dataAttr } from "@chakra-ui/utils"
import { chakra, ComponentWithProps, DeepPartial } from "@chakra-ui/vue-system"
import { extractImgAttrs, getInitials } from "./utils"
import {
  omitThemingProps,
  SystemStyleObject,
  ThemingProps,
} from "@chakra-ui/styled-system"
import {
  createStylesContext,
  DOMElements,
  useMultiStyleConfig,
} from "@chakra-ui/vue-system"
import { CAvatarImage } from "./c-avatar-image"
import { createIconComponent } from "@chakra-ui/c-icon"

export const baseStyle: SystemStyleObject = {
  display: "inline-flex",
  alignItems: "center",
  justifyContent: "center",
  textAlign: "center",
  textTransform: "uppercase",
  fontWeight: "medium",
  position: "relative",
  flexShrink: 0,
}

type BorderRadiusType = SystemStyleObject["borderRadius"]
type BorderColorType = SystemStyleObject["borderColor"]

export interface AvatarOptions {
  /**
   * The name of the person in the avatar.
   *
   * - if `src` has loaded, the name will be used as the `alt` attribute of the `img`
   * - If `src` is not loaded, the name will be used to create the initials
   */
  name?: string
  /**
   * If `true`, the `Avatar` will show a border around it.
   *
   * Best for a group of avatars
   *
   * @default false
   */
  showBorder?: boolean

  /**
   * The image url of the `Avatar`
   */
  src?: string
  /**
   * List of sources to use for different screen resolutions
   */
  srcset?: string
  /**
   * Defines loading strategy
   */
  loading?: "eager" | "lazy"
  /**
   * The border color of the avatar
   * @type SystemProps["borderColor"]
   */
  borderColor?: SystemStyleObject["borderColor"]

  /**
   * Initials override for the avatar
   */
  initials?: string
  /**
   * Defining which referrer is sent when fetching the resource.
   */
  referrerPolicy?: HTMLImageElement["referrerPolicy"]
}

const [AvatarStylesProvider, useAvatarStyles] = createStylesContext("Avatar")
export { AvatarStylesProvider, useAvatarStyles }

export interface CAvatarContext {
  icon: ComputedRef<VNode | undefined>
  isLoaded: ComputedRef<boolean>
}
const [AvatarContextProvider, useAvatarContext] = createContext<CAvatarContext>(
  {
    name: "AvatarContext",
    strict: true,
  }
)

export { AvatarContextProvider, useAvatarContext }

export interface CAvatarProps extends AvatarOptions, ThemingProps<"Avatar"> {
  /**
   * Icon name or component
   */
  icon?: string | DefineComponent
  iconLabel?: string
  /**
   * If `true`, opt out of the avatar's `fallback` logic and
   * renders the `img` at all times.
   *
   * @default false
   */
  ignoreFallback?: boolean
}

/**
 * CAvatar Component
 */
export const CAvatar = defineComponent({
  name: "CAvatar",
  props: {
    as: {
      type: [Object, String] as PropType<DOMElements>,
      default: "span",
    },
    src: {
      type: String as PropType<CAvatarProps["src"]>,
      default: "",
    },
    srcSet: {
      type: String as PropType<string>,
      default: "",
    },
    name: {
      type: String as PropType<CAvatarProps["name"]>,
      default: "",
    },
    initials: {
      type: String as PropType<CAvatarProps["initials"]>,
      default: "",
    },
    loading: String as PropType<CAvatarProps["loading"]>,
    showBorder: Boolean as PropType<CAvatarProps["showBorder"]>,
    iconLabel: String as PropType<CAvatarProps["iconLabel"]>,
    borderRadius: {
      type: SNAO as PropType<SystemStyleObject["borderRadius"]>,
      default: "full",
    },
    borderColor: SNAO as PropType<SystemStyleObject["borderColor"]>,
    ignoreFallback: SNAO as PropType<CAvatarProps["ignoreFallback"]>,
    icon: [String, Object] as PropType<CAvatarProps["icon"]>,
    ...vueThemingProps,
  },
  emits: ["load", "error"],
  setup(props, { slots, attrs, emit }) {
    const themingProps = useThemingProps(props)
    const mergedProps = computed(() => mergeWith({}, props, attrs))
    const ownProps = computed(() =>
      filterUndefined(omitThemingProps(mergedProps.value))
    )
    const extractedAttrs = computed(() => extractImgAttrs(ownProps.value))

    const isLoaded = ref(false)

    const styles = useMultiStyleConfig(
      "Avatar",
      computed(() => ({
        ...themingProps.value,
        name: props.name,
      }))
    )

    function handleLoaded(value: unknown) {
      if (value) {
        isLoaded.value = true
        emit("load", value)
      }
    }

    const ResolvedAvatarIcon = computed<VNode | undefined>(() => {
      // Here we prioritize the icon slot
      const iconSlot = slots.icon?.() as VNode[]
      if (iconSlot) {
        if (iconSlot) {
          const [firstChild] = iconSlot
          return firstChild
        }
      } else if (props.icon) {
        return match(typeof props.icon as any, {
          string: h(createIconComponent(props.icon as string) as any),
          object: h(props.icon as any),
        })
      }

      return undefined
    })

    const avatarStyles = computed<SystemStyleObject>(() => ({
      borderRadius: props.borderRadius,
      borderWidth: props.showBorder ? "2px" : undefined,
      ...baseStyle,
      ...styles.value.container,
      ...(props.borderColor && { borderColor: props.borderColor }),
    }))

    AvatarStylesProvider(styles)
    AvatarContextProvider({
      icon: ResolvedAvatarIcon,
      isLoaded: computed(() => isLoaded.value),
    })

    return () => (
      <chakra.span
        __label="avatar"
        data-loaded={dataAttr(isLoaded.value)}
        __css={avatarStyles.value}
        {...attrs}
      >
        <CAvatarImage
          src={props.src}
          srcSet={props.srcSet}
          loading={props.loading}
          onLoad={handleLoaded}
          onError={(e) => emit("error", e)}
          initials={props.initials}
          name={props.name}
          // @ts-ignore
          borderRadius={props.borderRadius}
          iconLabel={props.iconLabel}
          ignoreFallback={props.ignoreFallback}
          {...extractedAttrs.value.imgAttrs}
        />
        {slots?.default?.()}
      </chakra.span>
    )
  },
}) as any as ComponentWithProps<CAvatarProps>
