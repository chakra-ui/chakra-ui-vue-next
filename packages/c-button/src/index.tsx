import { defineComponent } from "vue"

export { default as CButton } from "./button"
export { default as CButtonGroup } from "./button-group"
export { default as CIconButton } from "./icon-button"
