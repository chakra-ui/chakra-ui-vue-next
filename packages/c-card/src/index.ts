export { CCard, type CCardProps } from "./c-card"
export { CCardBody, type CCardBodyProps } from "./c-card-body"
export { CCardHeader, type CCardHeaderProps } from "./c-card-header"
