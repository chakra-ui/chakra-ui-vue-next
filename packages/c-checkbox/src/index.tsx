export { CCheckbox } from "./checkbox"
export {
  CCheckboxGroup,
  useCheckboxGroupContext,
  type CheckboxGroupContext,
} from "./checkbox-group"

export type { CCheckboxControlProps, CCheckboxProps } from "./checkbox"
export type { CCheckboxGroupProps } from "./checkbox-group"
