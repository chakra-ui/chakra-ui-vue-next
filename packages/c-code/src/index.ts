export { default as CCode } from './code'
