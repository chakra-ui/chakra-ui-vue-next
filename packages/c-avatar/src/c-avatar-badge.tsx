import { SystemStyleObject } from "@chakra-ui/styled-system"
import { warn } from "@chakra-ui/utils"
import { chakra, HTMLChakraProps } from "@chakra-ui/vue-system"
import {
  computed,
  defineComponent,
  getCurrentInstance,
  PropType,
  h,
  Fragment,
} from "vue"
import { useAvatarStyles } from "./c-avatar"

type BadgePlacement = "top-start" | "top-end" | "bottom-start" | "bottom-end"

const placementMap: Record<BadgePlacement, SystemStyleObject> = {
  "top-start": {
    top: "0",
    insetStart: "0",
    transform: "translate(-25%, -25%)",
  },
  "top-end": {
    top: "0",
    insetEnd: "0",
    transform: "translate(25%, -25%)",
  },
  "bottom-start": {
    bottom: "0",
    insetStart: "0",
    transform: "translate(-25%, 25%)",
  },
  "bottom-end": {
    bottom: "0",
    insetEnd: "0",
    transform: "translate(25%, 25%)",
  },
}

export interface AvatarBadgeProps extends HTMLChakraProps<"div"> {
  placement?: BadgePlacement
}

/**
 * CAvatarBadge used to show extra badge to the top-right
 * or bottom-right corner of an avatar.
 */

export const CAvatarBadge = defineComponent({
  name: "CAvatarBadge",
  props: {
    placement: {
      type: String as PropType<BadgePlacement>,
      default: "bottom-end",
    },
  },
  setup(props, { attrs, slots }) {
    const styles = useAvatarStyles()

    const instance = getCurrentInstance()

    warn.bind(instance)({
      condition: !Object.keys(placementMap).includes(props.placement),
      message: `<CAvatarBadge /> expects a placement value of ${Object.keys(
        placementMap
      ).join(", ")}. Instead got "${props.placement}"`,
    })

    const badgeStyles = computed<SystemStyleObject>(() => ({
      position: "absolute",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      boxSize: "1em",
      ...(placementMap[props.placement] || {}),
      ...styles.value.badge,
    }))

    return () => (
      <chakra.div
        __label={"avatar__badge"}
        __css={badgeStyles.value}
        {...attrs}
      >
        {slots?.default?.()}
      </chakra.div>
    )
  },
})
