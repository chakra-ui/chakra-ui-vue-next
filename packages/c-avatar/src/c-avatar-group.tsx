/* eslint vue/no-side-effects-in-computed-properties: 0  */
import { useMultiStyleConfig, chakra } from "@chakra-ui/vue-system"
import { SystemStyleObject } from "@chakra-ui/styled-system"
import {
  getValidChildren,
  SNAO,
  vueThemingProps,
  useThemingProps,
  mergeWith,
} from "@chakra-ui/vue-utils"
import { filterUndefined } from "@chakra-ui/utils"

import {
  h,
  Fragment,
  defineComponent,
  computed,
  PropType,
  cloneVNode,
} from "vue"
import { baseStyle } from "./c-avatar"
import type * as CSS from "csstype"

export const avatarGroupProps = {
  max: {
    type: Number as PropType<number>,
    default: 2,
  },
  spacing: {
    type: SNAO as PropType<SystemStyleObject["margin"]>,
    default: "-0.75rem",
  },
  borderRadius: {
    type: SNAO as PropType<SystemStyleObject["borderRadius"]>,
    default: "full",
  },
  borderColor: SNAO as PropType<SystemStyleObject["borderColor"]>,
  ...vueThemingProps,
}

/**
 * CAvatarGroup displays a number of avatars grouped together in a stack.
 */
export const CAvatarGroup = defineComponent({
  name: "CAvatarGroup",
  props: avatarGroupProps,
  setup(props, { slots, attrs }) {
    const mergedProps = computed(() => mergeWith({}, props, attrs))
    const themingProps = useThemingProps(mergedProps.value)
    const styles = useMultiStyleConfig("Avatar", themingProps.value)

    const validChildren = computed(() => getValidChildren(slots))
    const visibleChildren = computed(() =>
      validChildren.value.slice(0, props.max)
    )
    const excessChildrenCount = computed(
      () => validChildren.value.length - props.max
    )

    /**
     * Reversing the children is a great way to avoid using zIndex
     * to overlap the avatars
     */
    const reversedVisibleChildren = computed(() =>
      visibleChildren.value.reverse()
    )

    const clonedChildren = computed(() =>
      reversedVisibleChildren.value.map((vnode, index) => {
        const isFirstAvatar = index === 0
        const childProps = filterUndefined({
          marginEnd: isFirstAvatar ? 0 : props.spacing,
          size: props.size,
          borderColor: vnode?.props?.borderColor ?? props.borderColor,
          showBorder: true,
        })
        return cloneVNode(vnode, childProps)
      })
    )

    const groupStyles = computed<SystemStyleObject>(() => ({
      display: "flex",
      alignItems: "center",
      justifyContent: "flex-end",
      flexDirection: "row-reverse",
      ...styles.value.group,
    }))

    const excessStyles = computed<SystemStyleObject>(() => ({
      borderRadius: props.borderRadius,
      marginStart: props.spacing,
      ...baseStyle,
      ...styles.value.excessLabel,
    }))

    return () => (
      <chakra.div
        role="group"
        __label="avatar__group"
        __css={groupStyles.value}
      >
        <>
          {excessChildrenCount.value > 0 && (
            <chakra.span __label="avatar__excess" __css={excessStyles.value}>
              {`+${excessChildrenCount.value}`}
            </chakra.span>
          )}
          {clonedChildren.value}
        </>
      </chakra.div>
    )
  },
})
