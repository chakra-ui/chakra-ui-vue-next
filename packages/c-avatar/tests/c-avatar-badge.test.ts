import { render, waitMs } from "@chakra-ui/vue-test-utils"
import { CAvatar, CAvatarBadge, avatarProps } from "../src"

const renderAvatar = (props?: any) => {
  const placement = props.placement ? `placement='${props.placement}'` : ""
  const template = `<CAvatar name='${props.name}' src='${props.src}' ><CAvatarBadge ${placement} /></CAvatar>`

  const component = {
    components: {
      CAvatar,
      CAvatarBadge,
    },
    template,
    props: avatarProps,
  }

  return render(component)
}

describe("CAvatarBadge", () => {
  it("renders badge without parameters", () => {
    const { asFragment } = renderAvatar({ name: "Evan You" })
    expect(asFragment()).toMatchSnapshot()
  })
})
