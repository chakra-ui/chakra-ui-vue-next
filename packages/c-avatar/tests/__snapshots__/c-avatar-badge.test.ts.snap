// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CAvatarBadge renders badge without parameters 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar chakra-1prtvd5"
  >
    <div
      aria-label="Evan You"
      class="chakra-uo84k3"
      role="img"
    >
      EY
    </div>
    <div
      class="chakra-avatar__badge chakra-173vitc"
    />
  </span>
</DocumentFragment>
`;
